### PR TITLE
Added gpio_shakti driver

### PR DESCRIPTION
**Pull request by Kishore2002**

### CODE REVIEW

File: drivers/gpio/Kconfig.shakti [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# SiFive Freedom GPIO configuration option
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/Kconfig.shakti:1]: Must be Secure-IoT GPIO ...

File: drivers/gpio/Kconfig.shakti [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+# Copyright (c) 2017 Jean-Paul Etienne <fractalclone@gmail.com>
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/Kconfig.shakti:3]: Copyright (c) 2024 Mindgrove Technologies Pvt Ltd.

File: drivers/gpio/gpio_shakti.c [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+#define DT_DRV_COMPAT shakti_gpio
+
+#ifdef CONFIG_BOARD_SECURE_IOT
+
+#define CLOCK_FREQUENCY_SIM 1000000000
+#define CLOCK_FREQUENCY 40000000
+#endif
+
+
+#define GPIO_START 0x00040200 //GPIO Start Address */
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:46]: Remove */ in the end of line 16.

File: drivers/gpio/gpio_shakti.c [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+
+#define GPIO_DIRECTION_CNTRL_REG  (GPIO_START + (0 * GPIO_OFFSET ))
+#define GPIO_DATA_REG             (GPIO_START + (1 * GPIO_OFFSET ))
+#define GPIO_SET_REG              (GPIO_START + (2 * GPIO_OFFSET ))
+#define GPIO_CLEAR_REG            (GPIO_START + (3 * GPIO_OFFSET ))
+#define GPIO_TOGGLE_REG           (GPIO_START + (4 * GPIO_OFFSET ))
+#define GPIO_QUAL_REG             (GPIO_START + (5 * GPIO_OFFSET ))
+#define GPIO_INTERRUPT_CONFIG_REG (GPIO_START + (6 * GPIO_OFFSET ))
+
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:57]: No Space before close parenthesis ')'. must be (x * GPIO_OFFSET).

File: drivers/gpio/gpio_shakti.c [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+/* Helper Macros for GPIO */
+#define DEV_GPIO_CFG(dev)						\
+	((const struct gpio_shakti_config * const)(dev)->config)
+#define DEV_GPIO(dev)							\
+	((volatile struct gpio_shakti_regs_t *)(DEV_GPIO_CFG(dev))->gpio_base_addr)
+#define DEV_GPIO_DATA(dev)				\
+	((struct gpio_shakti_data *)(dev)->data)
+
+// struct gpio_shakti_regs_t *gpio;
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:153]: remove

File: drivers/gpio/gpio_shakti.c [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+static struct gpio_shakti_data gpio_shakti_data0;
+
+DEVICE_DT_INST_DEFINE(0,
+                gpio_shakti_init,
+                NULL,
+                &gpio_shakti_data0, &gpio_shakti_config0,
+                PRE_KERNEL_1, CONFIG_GPIO_INIT_PRIORITY,
+                &gpio_shakti_driver);
+
+
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:330]: Why this much space below? Remove.

File: drivers/gpio/gpio_shakti.c [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+/*configuration function that configures the whether the gpio pin in input or output */
+
+static int gpio_shakti_pin_configure (const struct device *dev, 
+                        gpio_pin_t pin, 
+                        gpio_flags_t flags){
+
+    volatile struct gpio_shakti_regs_t *gpio = DEV_GPIO(dev);
+    // const struct gpio_shakti_regs_t *cfg = DEV_GPIO_CFG(dev);
+    printk("Configuration Done11111111111\n");
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:194]: Why so many 1's?

File: drivers/gpio/gpio_shakti.c [hunk 8]
```diff
@@ -1,3 +1,33 @@
+/****************************************************************************
+ * Project                               : Secure IoT SoC
+ * Name of the file                      : gpio_shakthi.c
+ * Brief Description of file             : This is a ZEPHYR OS GPIO Driver file for Mindgrove Silicon's GPIO Peripheral.
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:4]: The Description of the file must be inside the asterisks. Maximum length in a line is 100 columns.

File: drivers/gpio/gpio_shakti.c [hunk 9]
```diff
@@ -1,3 +1,33 @@
...
+ This program is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *****************************************************************************/
+/**
+  @file gpio_shakthi.c
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:25]: Change filename. gpio_shakti.c

File: drivers/gpio/gpio_shakti.c [hunk 10]
```diff
@@ -1,3 +1,33 @@
...
+  @brief Contains the driver routines for GPIO interface.
+  @details The GPIO driver has software functions to configure, set, clear
+  toggle signal over GPIO interface.
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:28]: No white space after fullstop.

File: drivers/gpio/gpio_shakti.c [hunk 11]
```diff
@@ -1,3 +1,33 @@
+/****************************************************************************
+ * Project                               : Secure IoT SoC
+ * Name of the file                      : gpio_shakthi.c
+ * Brief Description of file             : This is a ZEPHYR OS GPIO Driver file for Mindgrove Silicon's GPIO Peripheral.
+ * Name of Author                        : Kishore. J
+ * Email ID                              : <kishore@mindgrovetech.in>
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:6]: no white-space after full-stop or '>' etc

File: drivers/gpio/gpio_shakti.c [hunk 12]
```diff
@@ -123,19 +153,38 @@ struct gpio_shakti_data {
 // struct gpio_shakti_regs_t *gpio;
 
 /* Initialize's the gpio's*/
+/**
+ * The function initializes the GPIO device and prints a message indicating that initialization is
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:157]: @brief The function ...

File: drivers/gpio/gpio_shakti.c [hunk 13]
```diff
@@ -123,19 +153,38 @@ struct gpio_shakti_data {
 // struct gpio_shakti_regs_t *gpio;
 
 /* Initialize's the gpio's*/
+/**
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:156]: @fn int gpio_shakti_init(...)

File: drivers/gpio/gpio_shakti.c [hunk 14]
```diff
@@ -123,19 +153,38 @@ struct gpio_shakti_data {
...
+ * The function initializes the GPIO device and prints a message indicating that initialization is
+ * complete.
+ *
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:159]: @details ?

File: drivers/gpio/gpio_shakti.c [hunk 15]
```diff
@@ -123,19 +153,38 @@ struct gpio_shakti_data {
...
 
-    gpio = GPIO_START;
+    *gpio = GPIO_START;
     printk("Initialization Done\n");
 }
 
 
 /*configuration function that configures the whether the gpio pin in input or output */
 
+/**
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:176]: same as above

File: drivers/gpio/gpio_shakti.c [hunk 16]
```diff
@@ -152,6 +201,18 @@ static int gpio_shakti_pin_configure (const struct device *dev,
 
 /* Reads the data from the data register */
 
+/**
+ * The function gpio_shakti_pin_get_raw retrieves the raw value of a GPIO pin.
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:205]: same as above. Do this for all functions.

File: drivers/gpio/gpio_shakti.c [hunk 17]
```diff
@@ -219,8 +316,12 @@ static const struct gpio_shakti_config gpio_shakti_config0 ={
 
 static struct gpio_shakti_data gpio_shakti_data0;
 
-DEVICE_DT_INST_DEFINE(0,
+/* `DEVICE_DT_INST_DEFINE` is a macro provided by the Zephyr RTOS that is used to define a device
+instance. It is typically used in the board configuration file to define the devices present on the
+board. */
+
                 gpio_shakti_init,
```
Comment (M-Kapil-Shyam) [inline drivers/gpio/gpio_shakti.c:323]: why is gpio_shakti_init, outside the parenthesis?

File: drivers/gpio/Kconfig.shakti (hunk 1 above)
Comment (Kishore2002) [reply drivers/gpio/Kconfig.shakti]: Done

File: drivers/gpio/Kconfig.shakti (hunk 2 above)
Comment (Kishore2002) [reply drivers/gpio/Kconfig.shakti]: Done

File: drivers/gpio/gpio_shakti.c (hunk 3 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 4 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c:57]: Done

File: drivers/gpio/gpio_shakti.c [hunk 18]
```diff
@@ -0,0 +1,355 @@
...
+    uint32_t  clear;                   /*! clear register */
+    uint32_t  reserved3;                /*! reserved for future use */
+    uint32_t  toggle;               /*! toggle register */
+    uint32_t  reserved4;                /*! reserved for future use */
+    uint8_t  qualification;    /*! qualification register */
+    uint8_t  reserved5;                /*! reserved for future use */
+    uint16_t  reserved6;              /*! reserved for future use */
+    uint32_t  reserved12;              /*! reserved for future use */
+    uint32_t  intr_config;     /*! interrupt configuration register */
+//  uint16_t  reserved13;              /*! reserved for future use */
```
[the remark this answers is not in the document]
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 5 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 6 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 7 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 8 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c:4]: Done

File: drivers/gpio/gpio_shakti.c (hunk 9 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 10 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 11 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

File: drivers/gpio/gpio_shakti.c (hunk 17 above)
Comment (Kishore2002) [reply drivers/gpio/gpio_shakti.c]: Done

**Comment by M-Kapil-Shyam** [comment]
Complete the documentation.

**Pull request merged**